openlibrary/catalog/add_book/__init__.py

```diff
@@ -993,6 +1001,13 @@ def load(rec, account_key=None, from_marc_record: bool = False):
 
     normalize_import_record(rec)
 
+    # For recs with a non-ISBN ASIN, supplement the record with BookWorm data.
+    if non_isbn_asin := get_non_isbn_asin(rec):
+        from openlibrary.core.imports import ImportItem  # Evade circular import.
+
+        if item := ImportItem.find_staged_or_pending([non_isbn_asin]).first():
+            rec = json.loads(item.get("data", '{}')) | rec
```

Comment:
The PR title and description say "augment" and "supplement" but this looks like it simply replaces the record if a match is found. I don't know if it'll have much difference in practice, but it'll help to be clear about what is happening.  I don't think anyone has investigated whether the source Amazon record is guaranteed to be complete...
now that I think of it it _needs_ to be augmented because at the very least the BWB SKU and source record from the original record needs to be carried forward to enable the barcode matching we need for scanning. This might be a blocker.

Reply:
I may be repeating what you've just stated is the problem, so if that's the case, my apologies. I just want to get on the same page first.

What the highlighted code section is hopefully is doing is having the incoming `rec` data clobber fields that exist in both `rec` and whatever was found in the `import_item` table by way of `ImportItem.find_staged_or_pending()`.

E.g. in the following, the `publish_date` from `staged_import_item` is added to the returned dictionary, but the `title` is not:
 ```python
>>> staged_import_item = {"title": "A Great Book %!-", "identifiers": {"amazon": ["B012345678"]} , "publish_date": "2023"}
>>> rec = {"title": "A Great Book", "identifiers": {"amazon": ["B012345678"]}}
>>> staged_import_item | rec
{'title': 'A Great Book', 'identifiers': {'amazon': ['B012345678']}, 'publish_date': '2023'}
```

In this way, `rec` should be supplemented with whatever was found in the match from `import_item`, but no fields in `rec` should be overwritten.

Reply:
@scottbarnes That `|` is doing a lot of work, and it's hard to tell how the problem data is going to be processed, I don't have a good picture of what each of the two `rec`s have as keys.

Here are some real samples of promise item data from one of the promise items for testing:


    wget https://archive.org/download/bwb_daily_pallets_2023-11-02/DailyPallets__2023-11-02.json

First 10 no-Author no-Date B0* ASINs from promise item bwb_daily_pallets_2023-11-02 :

     sed 's/^\[//;s/},{/}\n{/g;s/\]$//'  DailyPallets__2023-11-02.json | grep B0 | grep '"Author":""' | grep '"PublicationDate":null' | head


```
{"BookBarcode":"KT-047-072","PackedLocation":"Dunfermline","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"KT-047-072","BookSKU":"","ISBN":"BWBM57920566","ASIN":"B0007IX3CM","ProductJSON":{"ISBN":"BWBM57920566","ASIN":"B0007IX3CM","Title":"The economics of \" under-developed \" areas;: An annotated reading list of books, articles, and official publications","MasterProductId":"57920566","BookId":"227351119","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"KT-017-548","PackedLocation":"Dunfermline","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"KT-017-548","BookSKU":"","ISBN":"BWBM57920580","ASIN":"B001W8W7EQ","ProductJSON":{"ISBN":"BWBM57920580","ASIN":"B001W8W7EQ","Title":"The New Cambridge Bibliography of English Literature","MasterProductId":"57920580","BookId":"227351595","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"KT-085-946","PackedLocation":"Dunfermline","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"KT-085-946","BookSKU":"","ISBN":"BWBM57920582","ASIN":"B000QV8BMM","ProductJSON":{"ISBN":"BWBM57920582","ASIN":"B000QV8BMM","Title":"The Barbarians: Warriors and Wars of the Dark Ages","MasterProductId":"57920582","BookId":"227351630","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-CCW-518","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-CCW-518","BookSKU":"","ISBN":"BWBM57920590","ASIN":"B000OSM4I4","ProductJSON":{"ISBN":"BWBM57920590","ASIN":"B000OSM4I4","Title":"The Next Move: Current Events in Bible Prophecy","MasterProductId":"57920590","BookId":"227351789","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-CCW-521","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-CCW-521","BookSKU":"","ISBN":"BWBM57920592","ASIN":"B0017XK6GW","ProductJSON":{"ISBN":"BWBM57920592","ASIN":"B0017XK6GW","Title":"The Experience","MasterProductId":"57920592","BookId":"227351786","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-CCW-524","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-CCW-524","BookSKU":"","ISBN":"BWBM57920595","ASIN":"B01JCR2WES","ProductJSON":{"ISBN":"BWBM57920595","ASIN":"B01JCR2WES","Title":"Nandi customary law","MasterProductId":"57920595","BookId":"227351962","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-CCW-551","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-CCW-551","BookSKU":"","ISBN":"BWBM57920622","ASIN":"B0000EE0HG","ProductJSON":{"ISBN":"BWBM57920622","ASIN":"B0000EE0HG","Title":"VISIONS OF AFRICA","MasterProductId":"57920622","BookId":"227352471","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-CCB-679","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-CCB-679","BookSKU":"","ISBN":"BWBM57920626","ASIN":"B004GF3UUW","ProductJSON":{"ISBN":"BWBM57920626","ASIN":"B004GF3UUW","Title":"Eusebius Werke, Achter Band: Die Praeparatio Evangelica, Erster Teil: Einleitung, Die Bucher I bis X (Die Griechischen Christlichen Schriftsteller der Ersten Jahrhunderte 8\/1)","MasterProductId":"57920626","BookId":"227352681","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-CCW-555","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-CCW-555","BookSKU":"","ISBN":"BWBM57920630","ASIN":"B0046GNRKO","ProductJSON":{"ISBN":"BWBM57920630","ASIN":"B0046GNRKO","Title":"TOPICS IN EAST AFRICAN HISTORY 1000-1970.","MasterProductId":"57920630","BookId":"227352779","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-CCB-681","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-CCB-681","BookSKU":"","ISBN":"BWBM57920633","ASIN":"B001KVTCLG","ProductJSON":{"ISBN":"BWBM57920633","ASIN":"B001KVTCLG","Title":"Hegemonius, Acta Archelai","MasterProductId":"57920633","BookId":"227352891","Author":"","Publisher":null,"PublicationDate":null}}
```

Random sample of 10:

    sed 's/^\[//;s/},{/}\n{/g;s/\]$//'  DailyPallets__2023-11-02.json | grep B0 | grep '"Author":""' | grep '"PublicationDate":null' | shuf -n10

```
{"BookBarcode":"P8-CQU-196","PackedLocation":"Reno","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"P8-CQU-196","BookSKU":"","ISBN":"BWBM57921210","ASIN":"B000YBOQM8","ProductJSON":{"ISBN":"BWBM57921210","ASIN":"B000YBOQM8","Title":"Revelations of an Opera Manager in 19th Century America","MasterProductId":"57921210","BookId":"227392511","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-BJH-243","PackedLocation":"York","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-BJH-243","BookSKU":"","ISBN":"BWBM57921011","ASIN":"B09CSCZ3RD","ProductJSON":{"ISBN":"BWBM57921011","ASIN":"B09CSCZ3RD","Title":"Letters from Madame la Marquise de Sevigne. Selected, translated, and introduced by Violet Hammersley. With a preface by W. Somerset Maugham","MasterProductId":"57921011","BookId":"227376486","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"P9-AHG-152","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"P9-AHG-152","BookSKU":"","ISBN":"BWBM57921312","ASIN":"B00KFT10UA","ProductJSON":{"ISBN":"BWBM57921312","ASIN":"B00KFT10UA","Title":"A Taste of Heaven - United Methodist Women Church Cookbook, North Carolina Cook Book","MasterProductId":"57921312","BookId":"227403938","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"P8-CQP-559","PackedLocation":"Reno","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"P8-CQP-559","BookSKU":"","ISBN":"BWBM57921089","ASIN":"B009NNX69W","ProductJSON":{"ISBN":"BWBM57921089","ASIN":"B009NNX69W","Title":"Baby Looney Tunes I Play","MasterProductId":"57921089","BookId":"227380126","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"P9-AYU-600","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"P9-AYU-600","BookSKU":"","ISBN":"BWBM57921206","ASIN":"B077KLGN71","ProductJSON":{"ISBN":"BWBM57921206","ASIN":"B077KLGN71","Title":"Born in Paradise","MasterProductId":"57921206","BookId":"227392442","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"P8-CRU-380","PackedLocation":"Reno","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"P8-CRU-380","BookSKU":"","ISBN":"BWBM57920894","ASIN":"B0006RX9JG","ProductJSON":{"ISBN":"BWBM57920894","ASIN":"B0006RX9JG","Title":"Sharing your faith with people of other faiths","MasterProductId":"57920894","BookId":"227372026","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-ATC-548","PackedLocation":"Reno","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-ATC-548","BookSKU":"","ISBN":"BWBM57920846","ASIN":"B001QVTZ1W","ProductJSON":{"ISBN":"BWBM57920846","ASIN":"B001QVTZ1W","Title":"Warbonnet Law (Signet Brand Western, 451-Q5867-095)","MasterProductId":"57920846","BookId":"227365450","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-ACL-477","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-ACL-477","BookSKU":"","ISBN":"BWBM57920754","ASIN":"B001P8E2R8","ProductJSON":{"ISBN":"BWBM57920754","ASIN":"B001P8E2R8","Title":"Pagan and Christian Rome","MasterProductId":"57920754","BookId":"227361420","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"KT-042-281","PackedLocation":"Dunfermline","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"KT-042-281","BookSKU":"","ISBN":"BWBM57920797","ASIN":"B000TLRD0U","ProductJSON":{"ISBN":"BWBM57920797","ASIN":"B000TLRD0U","Title":"Principles of Understanding: an Introduction to Logic from the Standpoint of Personal Idealism","MasterProductId":"57920797","BookId":"227363861","Author":"","Publisher":null,"PublicationDate":null}}
{"BookBarcode":"O9-CCW-604","PackedLocation":"Mishawaka","Sort":"Never Seen","PalletBarcode":"11\/1\/2023 8:30:14 PM","BookSKUB":"O9-CCW-604","BookSKU":"","ISBN":"BWBM57920699","ASIN":"B001AIR496","ProductJSON":{"ISBN":"BWBM57920699","ASIN":"B001AIR496","Title":"The Maasai","MasterProductId":"57920699","BookId":"227359752","Author":"","Publisher":null,"PublicationDate":null}}
```


I think it is scripts/promise_batch_imports.py that converts these to record JSON for import.

Reply:
@hornc, I am still fiddling with this, but here is how those 10 items might look.

Before:
```json
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B000YBOQM8"], "better_world_books": ["BWBM57921210"]}, "local_id": ["urn:bwbsku:P8-CQU-196"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:P8-CQU-196"], "title": "Revelations of an Opera Manager in 19th Century America"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B09CSCZ3RD"], "better_world_books": ["BWBM57921011"]}, "local_id": ["urn:bwbsku:O9-BJH-243"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:O9-BJH-243"], "title": "Letters from Madame la Marquise de Sevigne. Selected, translated, and introduced by Violet Hammersley. With a preface by W. Somerset Maugham"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B00KFT10UA"], "better_world_books": ["BWBM57921312"]}, "local_id": ["urn:bwbsku:P9-AHG-152"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:P9-AHG-152"], "title": "A Taste of Heaven - United Methodist Women Church Cookbook, North Carolina Cook Book"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B009NNX69W"], "better_world_books": ["BWBM57921089"]}, "local_id": ["urn:bwbsku:P8-CQP-559"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:P8-CQP-559"], "title": "Baby Looney Tunes I Play"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B077KLGN71"], "better_world_books": ["BWBM57921206"]}, "local_id": ["urn:bwbsku:P9-AYU-600"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:P9-AYU-600"], "title": "Born in Paradise"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B0006RX9JG"], "better_world_books": ["BWBM57920894"]}, "local_id": ["urn:bwbsku:P8-CRU-380"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:P8-CRU-380"], "title": "Sharing your faith with people of other faiths"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B001QVTZ1W"], "better_world_books": ["BWBM57920846"]}, "local_id": ["urn:bwbsku:O9-ATC-548"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:O9-ATC-548"], "title": "Warbonnet Law (Signet Brand Western, 451-Q5867-095)"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B001P8E2R8"], "better_world_books": ["BWBM57920754"]}, "local_id": ["urn:bwbsku:O9-ACL-477"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:O9-ACL-477"], "title": "Pagan and Christian Rome"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B000TLRD0U"], "better_world_books": ["BWBM57920797"]}, "local_id": ["urn:bwbsku:KT-042-281"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:KT-042-281"], "title": "Principles of Understanding: an Introduction to Logic from the Standpoint of Personal Idealism"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B001AIR496"], "better_world_books": ["BWBM57920699"]}, "local_id": ["urn:bwbsku:O9-CCW-604"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:O9-CCW-604"], "title": "The Maasai"}
```

After:
```json
{"authors": [{"name": "Max Maretzek"}, {"name": "Poesal"}], "identifiers": {"amazon": ["B000YBOQM8"], "better_world_books": ["BWBM57921210"]}, "local_id": ["urn:bwbsku:P8-CQU-196"], "physical_format": "paperback", "publish_date": "Jan 01, 1968", "publishers": ["DOVER-1957"], "source _records": ["promise:bwb_daily_pallets_2023-11-02:P8-CQU-196"], "title": "Revelations of an Opera Manager in 19th Century America"}
{"authors": [{"name": "Marie de Rabutin-Chantal d S_vign_"}], "identifiers": {"amazon": ["B09CSCZ3RD"], "better_world_books": ["BWBM57921011"]}, "local_id": ["urn:bwbsku:O9-BJH-243"], "physical_format": "hardcover", "publish_date": "Apr 23, 1956", "publishers": ["New York, Harcourt, Brace and Company"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:O9-BJH-243"], "title": "Letters from Madame la Marquise de Sevigne. Selected, translated, and introduced by Violet Hammersley. With a preface by W. Somerset Maugham"}
{"authors": [{"name": "Lebanon Methodist Women"}], "identifiers": {"amazon": ["B00KFT10UA"], "better_world_books": ["BWBM57921312"]}, "local_id": ["urn:bwbsku:P9-AHG-152"], "number_of_pages": 192, "physical_format": "loose leaf", "publish_date": "????", "publishers": ["Morris Press" ], "source_records": ["promise:bwb_daily_pallets_2023-11-02:P9-AHG-152"], "title": "A Taste of Heaven - United Methodist Women Church Cookbook, North Carolina Cook Book"}
{"authors": [{"name": "Editor"}], "identifiers": {"amazon": ["B009NNX69W"], "better_world_books": ["BWBM57921089"]}, "local_id": ["urn:bwbsku:P8-CQP-559"], "physical_format": "hardcover", "publish_date": "Apr 23, 2000", "publishers": ["DS-MAX"], "source_records": ["promise:bwb_daily _pallets_2023-11-02:P8-CQP-559"], "title": "Baby Looney Tunes I Play"}
{"authors": [{"name": "Tempski, Armine von"}], "identifiers": {"amazon": ["B077KLGN71"], "better_world_books": ["BWBM57921206"]}, "local_id": ["urn:bwbsku:P9-AYU-600"], "physical_format": "hardcover", "publish_date": "Apr 23, 1968", "publishers": ["Meredith Press"], "source_records" : ["promise:bwb_daily_pallets_2023-11-02:P9-AYU-600"], "title": "Born in Paradise"}
{"authors": [{"name": "Cooper, David C"}], "identifiers": {"amazon": ["B0006RX9JG"], "better_world_books": ["BWBM57920894"]}, "local_id": ["urn:bwbsku:P8-CRU-380"], "number_of_pages": 273, "physical_format": "unknown binding", "publish_date": "Apr 23, 1996", "publishers": ["David C. Cooper"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:P8-CRU-380"], "title": "Sharing your faith with people of other faiths"}
{"authors": [{"name": "Frank O'Rourke"}], "identifiers": {"amazon": ["B001QVTZ1W"], "better_world_books": ["BWBM57920846"]}, "local_id": ["urn:bwbsku:O9-ATC-548"], "physical_format": "mass market paperback", "publish_date": "Apr 23, 1967", "publishers": ["Signet"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:O9-ATC-548"], "title": "Warbonnet Law (Signet Brand Western, 451-Q5867-095)"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B001P8E2R8"], "better_world_books": ["BWBM57920754"]}, "local_id": ["urn:bwbsku:O9-ACL-477"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:O9-ACL-477"], "title": " Pagan and Christian Rome"}
{"authors": [{"name": "Henry Sturt"}], "identifiers": {"amazon": ["B000TLRD0U"], "better_world_books": ["BWBM57920797"]}, "local_id": ["urn:bwbsku:KT-042-281"], "physical_format": "hardcover", "publish_date": "Apr 23, 1915", "publishers": ["Cambridge University Press"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:KT-042-281"], "title": "Principles of Understanding: an Introduction to Logic from the Standpoint of Personal Idealism"}
{"authors": [{"name": "????"}], "identifiers": {"amazon": ["B001AIR496"], "better_world_books": ["BWBM57920699"]}, "local_id": ["urn:bwbsku:O9-CCW-604"], "publish_date": "????", "publishers": ["????"], "source_records": ["promise:bwb_daily_pallets_2023-11-02:O9-CCW-604"], "title": "The Maasai"}
```

Summary of changes:
- 7 books get 'good' `authors` data (and an eighth book gets `Editor` as the author).
- 7 books get 'good' `publish_date` data.
- 7 (or 8) books get 'good' `publishers` data.
- A handful of books get some other stuff, such as `number_of_pages` and `physical_format`.